src/server/frontend_wayland/idle_inhibit_v1.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "idle_inhibit_v1.h"
+
+#include "mir/executor.h"
+#include "mir/scene/surface.h"
+
+#include "wl_surface.h"
+
+#include <boost/throw_exception.hpp>
+#include <memory>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace ms = mir::scene;
+
+struct IdleInhibitV1Ctx
+{
+    std::shared_ptr<mir::Executor> const wayland_executor;
+    std::shared_ptr<ms::IdleHub> const idle_hub;
+};
+
+class IdleInhibitManagerV1Global : public mw::IdleInhibitManagerV1::Global
+{
+public:
+    IdleInhibitManagerV1Global(wl_display* display, std::shared_ptr<IdleInhibitV1Ctx> ctx);
+
+private:
+    void bind(wl_resource* new_resource) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitManagerV1 : public mw::IdleInhibitManagerV1
+{
+public:
+    IdleInhibitManagerV1(wl_resource* resource, std::shared_ptr<IdleInhibitV1Ctx>  ctx);
+
+private:
+    void create_inhibitor(struct wl_resource* id, struct wl_resource* surface) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitorV1 : public mw::IdleInhibitorV1
+{
+public:
+    IdleInhibitorV1(wl_resource *resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+    ~IdleInhibitorV1();
+
+private:
+    struct StateObserver : ms::IdleStateObserver
+    {
+        StateObserver()
+        {
+        }
+
+        void idle() override;
+
+        void active() override;
+    };
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+    std::shared_ptr<StateObserver> const state_observer;
```

Comment:
This doesn't do anything, does it? Previously it was inadvertently causing something to work, but I think it can be removed now.

Reply:
It works without it now! Still don't exactly understand why this was needed for it work under the old implementation, but it passes tests and runs fine without the StateObserver now.